File: pkg/service/config/backup_service.go
/*
 * @Description: 系统设置（数据库配置）备份服务
 * @Author: 安知鱼
 * @Date: 2025-10-19
 */
package config

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// BackupInfo 定义了备份的元数据信息
type BackupInfo struct {
	Filename    string    `json:"filename"`    // 备份文件名
	Size        int64     `json:"size"`        // 文件大小（字节）
	CreatedAt   time.Time `json:"created_at"`  // 创建时间
	Description string    `json:"description"` // 备份描述
	IsAuto      bool      `json:"is_auto"`     // 是否自动备份
}

// BackupService 定义了系统设置备份服务的接口
type BackupService interface {
	// CreateBackup 创建系统设置备份（导出数据库配置到备份文件）
	CreateBackup(ctx context.Context, description string, isAuto bool) (*BackupInfo, error)

	// ListBackups 列出所有备份
	ListBackups(ctx context.Context) ([]*BackupInfo, error)

	// RestoreBackup 从备份恢复系统设置（将备份文件导入数据库）
	RestoreBackup(ctx context.Context, filename string) error

	// DeleteBackup 删除指定的备份
	DeleteBackup(ctx context.Context, filename string) error

	// CleanOldBackups 清理旧的备份文件，保留最近的 keepCount 个
	CleanOldBackups(ctx context.Context, keepCount int) error

	// SetMaxBackupCount 设置最大备份数量
	SetMaxBackupCount(maxCount int) error

	// GetMaxBackupCount 获取最大备份数量
	GetMaxBackupCount() int
}

// backupService 是 BackupService 接口的实现
// 备份内容为「系统设置」即数据库中的配置表数据，与「导出配置」一致
type backupService struct {
	backupDir       string               // 备份目录
	importExportSvc ImportExportService // 用于导出/导入数据库配置
	maxBackupCount  int                  // 最大备份数量，0表示无限制
}

const backupFilePrefix = "settings_backup_"
const backupFileSuffix = ".json"

// NewBackupService 创建一个新的系统设置备份服务实例
func NewBackupService(backupDir string, importExportSvc ImportExportService) BackupService {
	return NewBackupServiceWithLimit(backupDir, importExportSvc, 10)
}

// NewBackupServiceWithLimit 创建一个新的系统设置备份服务实例，可指定最大备份数量
func NewBackupServiceWithLimit(backupDir string, importExportSvc ImportExportService, maxBackupCount int) BackupService {
	if err := os.MkdirAll(backupDir, 0755); err != nil {
		log.Printf("警告: 创建备份目录失败: %v", err)
	}
	if maxBackupCount < 0 {
		log.Printf("警告: 最大备份数量不能为负数，设置为默认值10")
		maxBackupCount = 10
	}
	return &backupService{
		backupDir:       backupDir,
		importExportSvc: importExportSvc,
		maxBackupCount:  maxBackupCount,
	}
}

// CreateBackup 创建系统设置备份：从数据库导出配置并写入备份文件
func (s *backupService) CreateBackup(ctx context.Context, description string, isAuto bool) (*BackupInfo, error) {
	data, err := s.importExportSvc.ExportConfig(ctx)
	if err != nil {
		return nil, fmt.Errorf("导出系统配置失败: %w", err)
	}

	timestamp := time.Now().Format("20060102_150405")
	backupFilename := backupFilePrefix + timestamp + backupFileSuffix
	backupPath := filepath.Join(s.backupDir, backupFilename)

	if err := os.WriteFile(backupPath, data, 0644); err != nil {
		return nil, fmt.Errorf("写入备份文件失败: %w", err)
	}

	metadata := BackupInfo{
		Filename:    backupFilename,
		Size:        int64(len(data)),
		CreatedAt:   time.Now(),
		Description: description,
		IsAuto:      isAuto,
	}

	if err := s.saveMetadata(backupFilename, &metadata); err != nil {
		log.Printf("警告: 保存备份元数据失败: %v", err)
	}

	log.Printf("✅ 系统设置备份成功: %s (共 %d 字节)", backupFilename, metadata.Size)

	if s.maxBackupCount > 0 {
		if err := s.CleanOldBackups(ctx, s.maxBackupCount); err != nil {
			log.Printf("警告: 自动清理旧备份失败: %v", err)
		}
	}

	return &metadata, nil
}

// ListBackups 列出所有系统设置备份（仅 .json 备份文件）
func (s *backupService) ListBackups(ctx context.Context) ([]*BackupInfo, error) {
	entries, err := os.ReadDir(s.backupDir)
	if err != nil {
		if os.IsNotExist(err) {
			return []*BackupInfo{}, nil
		}
		return nil, fmt.Errorf("读取备份目录失败: %w", err)
	}

	var backups []*BackupInfo
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		// 只处理系统设置备份文件，排除元数据文件
		if !strings.HasPrefix(name, backupFilePrefix) || !strings.HasSuffix(name, backupFileSuffix) {
			continue
		}
		if strings.HasSuffix(name, ".meta.json") {
			continue
		}

		metadata := s.loadMetadata(name)
		if metadata == nil {
			info, err := entry.Info()
			if err != nil {
				continue
			}
			metadata = &BackupInfo{
				Filename:    name,
				Size:        info.Size(),
				CreatedAt:   info.ModTime(),
				Description: "旧版本备份",
				IsAuto:      false,
			}
		}
		backups = append(backups, metadata)
	}

	sort.Slice(backups, func(i, j int) bool {
		return backups[i].CreatedAt.After(backups[j].CreatedAt)
	})

	return backups, nil
}

// RestoreBackup 从备份恢复系统设置：读取备份文件并导入到数据库
func (s *backupService) RestoreBackup(ctx context.Context, filename string) error {
	if err := s.validateBackupFilename(filename); err != nil {
		return err
	}

	backupPath := filepath.Join(s.backupDir, filename)
	if _, err := os.Stat(backupPath); err != nil {
		if os.IsNotExist(err) {
			return fmt.Errorf("备份文件不存在: %s", filename)
		}
		return fmt.Errorf("访问备份文件失败: %w", err)
	}

	// 恢复前先备份当前配置
	_, _ = s.CreateBackup(ctx, "恢复前自动备份", true)

	data, err := os.ReadFile(backupPath)
	if err != nil {
		return fmt.Errorf("读取备份文件失败: %w", err)
	}

	if err := s.importExportSvc.ImportConfig(ctx, bytes.NewReader(data)); err != nil {
		return fmt.Errorf("恢复系统设置失败: %w", err)
	}

	log.Printf("✅ 系统设置已从备份恢复: %s", filename)
	return nil
}

// DeleteBackup 删除指定的备份文件及其元数据
func (s *backupService) DeleteBackup(ctx context.Context, filename string) error {
	if err := s.validateBackupFilename(filename); err != nil {
		return err
	}
	backupPath := filepath.Join(s.backupDir, filename)
	if _, err := os.Stat(backupPath); err != nil {
		if os.IsNotExist(err) {
			return fmt.Errorf("备份文件不存在: %s", filename)
		}
		return fmt.Errorf("访问备份文件失败: %w", err)
	}
	if err := os.Remove(backupPath); err != nil {
		return fmt.Errorf("删除备份文件失败: %w", err)
	}
	metaPath := s.getMetadataPath(filename)
	if _, err := os.Stat(metaPath); err == nil {
		_ = os.Remove(metaPath)
	}
	log.Printf("✅ 备份已删除: %s", filename)
	return nil
}

const maxKeepCount = 100

// CleanOldBackups 清理旧备份，保留最近 keepCount 个
func (s *backupService) CleanOldBackups(ctx context.Context, keepCount int) error {
	if keepCount < 1 {
		return fmt.Errorf("保留数量必须大于0")
	}
	if keepCount > maxKeepCount {
		return fmt.Errorf("保留数量不能超过 %d", maxKeepCount)
	}
	backups, err := s.ListBackups(ctx)
	if err != nil {
		return err
	}
	if len(backups) <= keepCount {
		return nil
	}
	for i := keepCount; i < len(backups); i++ {
		_ = s.DeleteBackup(ctx, backups[i].Filename)
	}
	return nil
}

// validateBackupFilename 校验备份文件名，防止路径穿越或非法文件名
// 合法格式：settings_backup_YYYYMMDD_HHMMSS.json（时间戳固定 15 位）
func (s *backupService) validateBackupFilename(filename string) error {
	if filename == "" {
		return fmt.Errorf("备份文件名为空")
	}
	if strings.Contains(filename, "..") || strings.ContainsRune(filename, os.PathSeparator) ||
		strings.Contains(filename, "/") || strings.Contains(filename, "\\") {
		return fmt.Errorf("无效的备份文件名: %s", filename)
	}
	expectedLen := len(backupFilePrefix) + 15 + len(backupFileSuffix) // 16+15+5=36
	if len(filename) != expectedLen ||
		!strings.HasPrefix(filename, backupFilePrefix) ||
		!strings.HasSuffix(filename, backupFileSuffix) {
		return fmt.Errorf("无效的备份文件名: %s", filename)
	}
	// 中间时间戳必须为 20060102_150405 格式（仅数字与下划线）
	mid := filename[len(backupFilePrefix) : len(filename)-len(backupFileSuffix)]
	for _, r := range mid {
		if r != '_' && (r < '0' || r > '9') {
			return fmt.Errorf("无效的备份文件名: %s", filename)
		}
	}
	return nil
}

// getMetadataPath 获取元数据文件路径（备份文件 x.json 对应 x.meta.json）
func (s *backupService) getMetadataPath(filename string) string {
	base := strings.TrimSuffix(filename, backupFileSuffix)
	return filepath.Join(s.backupDir, base+".meta.json")
}

func (s *backupService) saveMetadata(filename string, metadata *BackupInfo) error {
	data, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.getMetadataPath(filename), data, 0644)
}

func (s *backupService) loadMetadata(filename string) *BackupInfo {
	data, err := os.ReadFile(s.getMetadataPath(filename))
	if err != nil {
		return nil
	}
	var metadata BackupInfo
	if err := json.Unmarshal(data, &metadata); err != nil {
		return nil
	}
	return &metadata
}

// SetMaxBackupCount 设置最大备份数量
func (s *backupService) SetMaxBackupCount(maxCount int) error {
	if maxCount < 0 {
		return fmt.Errorf("最大备份数量不能为负数")
	}
	s.maxBackupCount = maxCount
	if maxCount > 0 {
		ctx := context.Background()
		_ = s.CleanOldBackups(ctx, maxCount)
	}
	return nil
}

// GetMaxBackupCount 获取最大备份数量
func (s *backupService) GetMaxBackupCount() int {
	return s.maxBackupCount
}

File: server/src/backup/backup.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { SettingsService } from '../settings/settings.service';
import * as fs from 'fs';
import * as path from 'path';

/**
 * BackupInfo — metadata for a settings backup file.
 * Matches Go BackupInfo (backup_service.go).
 */
export interface BackupInfo {
  filename: string;
  size: number;
  created_at: string;
  description: string;
  is_auto: boolean;
}

const BACKUP_FILE_PREFIX = 'settings_backup_';
const BACKUP_FILE_SUFFIX = '.json';
const MAX_KEEP_COUNT = 100;
const DEFAULT_MAX_BACKUP_COUNT = 10;

/**
 * BackupService — CRUD for settings backup files.
 * Matches Go BackupService (backup_service.go).
 *
 * - createBackup: export settings, write JSON, save metadata
 * - listBackups: read backup dir, load metadata, sort by date desc
 * - restoreBackup: validate filename, create pre-restore backup, import settings
 * - deleteBackup: validate filename, delete file + metadata
 * - cleanOldBackups: delete oldest beyond keepCount
 */
@Injectable()
export class BackupService {
  private readonly logger = new Logger(BackupService.name);
  private readonly backupDir: string;
  private maxBackupCount: number;

  constructor(private readonly settingsService: SettingsService) {
    this.backupDir = path.join(process.cwd(), 'data', 'backups');
    this.maxBackupCount = DEFAULT_MAX_BACKUP_COUNT;

    // Ensure backup directory exists
    try {
      if (!fs.existsSync(this.backupDir)) {
        fs.mkdirSync(this.backupDir, { recursive: true });
      }
    } catch (err) {
      this.logger.warn(`创建备份目录失败: ${err}`);
    }
  }

  /**
   * CreateBackup — export current settings and write as backup file.
   * Matches Go backupService.CreateBackup (backup_service.go lines 87-122).
   */
  async createBackup(description: string, isAuto: boolean): Promise<BackupInfo> {
    const data = await this.settingsService.exportAll();
    const jsonStr = JSON.stringify(data, null, 2);
    const buffer = Buffer.from(jsonStr, 'utf-8');

    const timestamp = this.formatTimestamp(new Date());
    const backupFilename = BACKUP_FILE_PREFIX + timestamp + BACKUP_FILE_SUFFIX;
    const backupPath = path.join(this.backupDir, backupFilename);

    fs.writeFileSync(backupPath, jsonStr, 'utf-8');

    const metadata: BackupInfo = {
      filename: backupFilename,
      size: buffer.length,
      created_at: new Date().toISOString(),
      description,
      is_auto: isAuto,
    };

    try {
      this.saveMetadata(backupFilename, metadata);
    } catch (err) {
      this.logger.warn(`保存备份元数据失败: ${err}`);
    }

    this.logger.log(`系统设置备份成功: ${backupFilename} (共 ${metadata.size} 字节)`);

    // Auto-cleanup if maxBackupCount is set
    if (this.maxBackupCount > 0) {
      try {
        await this.cleanOldBackups(this.maxBackupCount);
      } catch (err) {
        this.logger.warn(`自动清理旧备份失败: ${err}`);
      }
    }

    return metadata;
  }

  /**
   * ListBackups — list all backup files sorted by date descending.
   * Matches Go backupService.ListBackups (backup_service.go lines 125-170).
   */
  async listBackups(): Promise<BackupInfo[]> {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(this.backupDir, { withFileTypes: true });
    } catch (err: any) {
      if (err.code === 'ENOENT') {
        return [];
      }
      throw new Error(`读取备份目录失败: ${err}`);
    }

    const backups: BackupInfo[] = [];
    for (const entry of entries) {
      if (entry.isDirectory()) continue;
      const name = entry.name;

      // Only process settings backup files, exclude metadata files
      if (!name.startsWith(BACKUP_FILE_PREFIX) || !name.endsWith(BACKUP_FILE_SUFFIX)) continue;
      if (name.endsWith('.meta.json')) continue;

      let metadata = this.loadMetadata(name);
      if (!metadata) {
        // Fallback to file stats if metadata is missing
        try {
          const stat = fs.statSync(path.join(this.backupDir, name));
          metadata = {
            filename: name,
            size: stat.size,
            created_at: stat.mtime.toISOString(),
            description: '旧版本备份',
            is_auto: false,
          };
        } catch {
          continue;
        }
      }
      backups.push(metadata);
    }

    // Sort by created_at descending (newest first)
    backups.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());

    return backups;
  }

  /**
   * RestoreBackup — restore settings from a backup file.
   * Matches Go backupService.RestoreBackup (backup_service.go lines 173-200).
   * Creates a pre-restore automatic backup before importing.
   */
  async restoreBackup(filename: string): Promise<void> {
    this.validateBackupFilename(filename);

    const backupPath = path.join(this.backupDir, filename);
    if (!fs.existsSync(backupPath)) {
      throw new Error(`备份文件不存在: ${filename}`);
    }

    // Create pre-restore backup
    try {
      await this.createBackup('恢复前自动备份', true);
    } catch (err) {
      this.logger.warn(`恢复前自动备份失败: ${err}`);
    }

    const content = fs.readFileSync(backupPath, 'utf-8');
    const data = JSON.parse(content);

    await this.settingsService.importAll(data);

    this.logger.log(`系统设置已从备份恢复: ${filename}`);
  }

  /**
   * DeleteBackup — delete a backup file and its metadata.
   * Matches Go backupService.DeleteBackup (backup_service.go lines 203-223).
   */
  async deleteBackup(filename: string): Promise<void> {
    this.validateBackupFilename(filename);

    const backupPath = path.join(this.backupDir, filename);
    if (!fs.existsSync(backupPath)) {
      throw new Error(`备份文件不存在: ${filename}`);
    }

    fs.unlinkSync(backupPath);

    // Delete metadata if it exists
    const metaPath = this.getMetadataPath(filename);
    if (fs.existsSync(metaPath)) {
      try {
        fs.unlinkSync(metaPath);
      } catch {
        // Ignore metadata deletion errors
      }
    }

    this.logger.log(`备份已删除: ${filename}`);
  }

  /**
   * CleanOldBackups — delete oldest backups beyond keepCount.
   * Matches Go backupService.CleanOldBackups (backup_service.go lines 228-246).
   */
  async cleanOldBackups(keepCount: number): Promise<void> {
    if (keepCount < 1) {
      throw new Error('保留数量必须大于0');
    }
    if (keepCount > MAX_KEEP_COUNT) {
      throw new Error(`保留数量不能超过 ${MAX_KEEP_COUNT}`);
    }

    const backups = await this.listBackups();
    if (backups.length <= keepCount) return;

    for (let i = keepCount; i < backups.length; i++) {
      try {
        await this.deleteBackup(backups[i].filename);
      } catch {
        // Continue deleting others even if one fails
      }
    }
  }

  /**
   * SetMaxBackupCount — update max backup count and auto-cleanup.
   * Matches Go backupService.SetMaxBackupCount (backup_service.go lines 301-310).
   */
  setMaxBackupCount(maxCount: number): void {
    if (maxCount < 0) {
      this.logger.warn('最大备份数量不能为负数，设置为默认值10');
      maxCount = DEFAULT_MAX_BACKUP_COUNT;
    }
    this.maxBackupCount = maxCount;
    if (maxCount > 0) {
      this.cleanOldBackups(maxCount).catch(() => {});
    }
  }

  getMaxBackupCount(): number {
    return this.maxBackupCount;
  }

  /**
   * validateBackupFilename — prevent path traversal, enforce format.
   * Matches Go backupService.validateBackupFilename (backup_service.go lines 250-272).
   * Valid format: settings_backup_YYYYMMDD_HHMMSS.json (15-char timestamp)
   */
  private validateBackupFilename(filename: string): void {
    if (!filename) {
      throw new Error('备份文件名为空');
    }
    // Prevent path traversal
    if (filename.includes('..') || filename.includes('/') || filename.includes('\\')) {
      throw new Error(`无效的备份文件名: ${filename}`);
    }
    // Validate format: prefix + 15-char timestamp + suffix
    const expectedLen = BACKUP_FILE_PREFIX.length + 15 + BACKUP_FILE_SUFFIX.length; // 16+15+5=36
    if (
      filename.length !== expectedLen ||
      !filename.startsWith(BACKUP_FILE_PREFIX) ||
      !filename.endsWith(BACKUP_FILE_SUFFIX)
    ) {
      throw new Error(`无效的备份文件名: ${filename}`);
    }
    // Timestamp part must be digits and underscores only
    const mid = filename.slice(BACKUP_FILE_PREFIX.length, filename.length - BACKUP_FILE_SUFFIX.length);
    for (const ch of mid) {
      if (ch !== '_' && (ch < '0' || ch > '9')) {
        throw new Error(`无效的备份文件名: ${filename}`);
      }
    }
  }

  /**
   * getMetadataPath — get metadata file path for a backup.
   * x.json -> x.meta.json
   * Matches Go getMetadataPath (backup_service.go lines 275-278).
   */
  private getMetadataPath(filename: string): string {
    const base = filename.slice(0, -BACKUP_FILE_SUFFIX.length);
    return path.join(this.backupDir, base + '.meta.json');
  }

  /**
   * saveMetadata — write metadata companion file.
   * Matches Go saveMetadata (backup_service.go lines 280-286).
   */
  private saveMetadata(filename: string, metadata: BackupInfo): void {
    const metaPath = this.getMetadataPath(filename);
    fs.writeFileSync(metaPath, JSON.stringify(metadata, null, 2), 'utf-8');
  }

  /**
   * loadMetadata — read metadata companion file.
   * Matches Go loadMetadata (backup_service.go lines 288-298).
   * Returns null if metadata file doesn't exist or is invalid.
   */
  private loadMetadata(filename: string): BackupInfo | null {
    const metaPath = this.getMetadataPath(filename);
    try {
      const content = fs.readFileSync(metaPath, 'utf-8');
      return JSON.parse(content) as BackupInfo;
    } catch {
      return null;
    }
  }

  /**
   * formatTimestamp — format Date as YYYYMMDD_HHMMSS (local time, matches Go).
   * Go uses time.Now().Format("20060102_150405") which is local time.
   */
  private formatTimestamp(date: Date): string {
    const pad = (n: number) => n.toString().padStart(2, '0');
    const year = date.getFullYear();
    const month = pad(date.getMonth() + 1);
    const day = pad(date.getDate());
    const hour = pad(date.getHours());
    const minute = pad(date.getMinutes());
    const second = pad(date.getSeconds());
    return `${year}${month}${day}_${hour}${minute}${second}`;
  }
}
